File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.java
class Solution {

    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        PriorityQueue<Integer> q = new PriorityQueue<>(new Comparator<Integer>() {
            @Override
            public int compare(Integer a, Integer b) {
                return -Integer.compare(a, b);
            }
        });
        for (int i = 1; i < heights.length; i++) {
            if (heights[i] > heights[i-1]) {
                int diff = heights[i] - heights[i-1];
                if (bricks >= diff) {
                    bricks -= diff;
                    q.add(diff);
                }
                else if (ladders > 0) {
                    if (q.size() > 0 && diff < q.peek()) {
                        bricks += q.poll()-diff;
                        ladders--;
                        q.add(diff);
                    }
                    else {
                        ladders--;
                    }
                }
                else {
                    return i-1;
                }
            }
        }
        return heights.length-1;
    }
}

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        n = len(heights)
        q = []
        for i in range(1, n):
            climb = heights[i] - heights[i-1]
            if climb <= 0:
                continue
            elif bricks >= climb:
                bricks -= climb
                heappush(q, -climb)
            elif ladders > 0:
                if len(q) > 0 and climb < -q[0]:
                    bricks += -heappop(q) - climb
                    heappush(q, -climb)
                ladders -= 1
            else:
                return i - 1
        return n - 1